File: 3387.maximize-amount-after-two-days-of-conversions.java
#
# @lc app=leetcode id=3387 lang=java
#
# [3387] Maximize Amount After Two Days of Conversions
#

# @lc code=start
class Solution {
    public double maxAmount(String initialCurrency, List<List<String>> pairs1, double[] rates1, List<List<String>> pairs2, double[] rates2) {
        Map<String, Map<String, Double>> graph1 = new HashMap<>();
        for (int i = 0; i < pairs1.size(); i++) {
            String a = pairs1.get(i).get(0);
            String b = pairs1.get(i).get(1);
            double r = rates1[i];
            graph1.computeIfAbsent(a, k -> new HashMap<>()).put(b, r);
            graph1.computeIfAbsent(b, k -> new HashMap<>()).put(a, 1.0 / r);
        }
        Map<String, Map<String, Double>> graph2 = new HashMap<>();
        for (int i = 0; i < pairs2.size(); i++) {
            String a = pairs2.get(i).get(0);
            String b = pairs2.get(i).get(1);
            double r = rates2[i];
            graph2.computeIfAbsent(a, k -> new HashMap<>()).put(b, r);
            graph2.computeIfAbsent(b, k -> new HashMap<>()).put(a, 1.0 / r);
        }
        // Compute mult1: initial -> C on day1
        Map<String, Double> mult1 = new HashMap<>();
        mult1.put(initialCurrency, 1.0);
        Set<String> visited1 = new HashSet<>();
        visited1.add(initialCurrency);
        Queue<String> q1 = new LinkedList<>();
        q1.offer(initialCurrency);
        while (!q1.isEmpty()) {
            String curr = q1.poll();
            Map<String, Double> neis = graph1.getOrDefault(curr, new HashMap<>());
            for (Map.Entry<String, Double> entry : neis.entrySet()) {
                String nei = entry.getKey();
                double rate = entry.getValue();
                if (!visited1.contains(nei)) {
                    visited1.add(nei);
                    mult1.put(nei, mult1.get(curr) * rate);
                    q1.offer(nei);
                }
            }
        }
        // Build revGraph2 for day2
        Map<String, Map<String, Double>> revGraph2 = new HashMap<>();
        for (Map.Entry<String, Map<String, Double>> outer : graph2.entrySet()) {
            String a = outer.getKey();
            for (Map.Entry<String, Double> inner : outer.getValue().entrySet()) {
                String b = inner.getKey();
                double r_ab = inner.getValue();
                revGraph2.computeIfAbsent(b, k -> new HashMap<>()).put(a, 1.0 / r_ab);
            }
        }
        // Compute mult_rev: initial -> C in revGraph2 (1 / (C -> initial in graph2))
        Map<String, Double> mult_rev = new HashMap<>();
        mult_rev.put(initialCurrency, 1.0);
        Set<String> visited_rev = new HashSet<>();
        visited_rev.add(initialCurrency);
        Queue<String> q_rev = new LinkedList<>();
        q_rev.offer(initialCurrency);
        while (!q_rev.isEmpty()) {
            String curr = q_rev.poll();
            Map<String, Double> neis = revGraph2.getOrDefault(curr, new HashMap<>());
            for (Map.Entry<String, Double> entry : neis.entrySet()) {
                String nei = entry.getKey();
                double rate = entry.getValue();
                if (!visited_rev.contains(nei)) {
                    visited_rev.add(nei);
                    mult_rev.put(nei, mult_rev.get(curr) * rate);
                    q_rev.offer(nei);
                }
            }
        }
        // mult2: C -> initial on day2
        Map<String, Double> mult2 = new HashMap<>();
        for (Map.Entry<String, Double> e : mult_rev.entrySet()) {
            mult2.put(e.getKey(), 1.0 / e.getValue());
        }
        // Max over intermediate C
        double ans = 1.0;
        for (Map.Entry<String, Double> e : mult1.entrySet()) {
            String c = e.getKey();
            double m1v = e.getValue();
            Double m2v = mult2.get(c);
            if (m2v != null) {
                ans = Math.max(ans, m1v * m2v);
            }
        }
        return ans;
    }
}
# @lc code=end

File: 3387.maximize-amount-after-two-days-of-conversions.go
#
# @lc app=leetcode id=3387 lang=golang
#
# [3387] Maximize Amount After Two Days of Conversions
#

# @lc code=start
func maxAmount(initialCurrency string, pairs1 [][]string, rates1 []float64, pairs2 [][]string, rates2 []float64) float64 {
		currencies := make(map[string]bool)
		currencies[initialCurrency] = true
		for _, p := range pairs1 {
			currencies[p[0]] = true
			currencies[p[1]] = true
		}
		for _, p := range pairs2 {
			currencies[p[0]] = true
			currencies[p[1]] = true
		}
		currlist := make([]string, 0, len(currencies))
		for c := range currencies {
			currlist = append(currlist, c)
		}
		N := len(currlist)
		idmap := make(map[string]int, N)
		for i, c := range currlist {
			idmap[c] = i
		}
		sid := idmap[initialCurrency]

		maxf := func(a, b float64) float64 {
			if a > b {
				return a
			}
			return b
		}

		// day1
		dist1 := make([][]float64, N)
		for i := 0; i < N; i++ {
			dist1[i] = make([]float64, N)
			for j := 0; j < N; j++ {
				if i == j {
					dist1[i][j] = 1.0
				} else {
					dist1[i][j] = 0.0
				}
			}
		}
		for idx := range pairs1 {
			a := idmap[pairs1[idx][0]]
			b := idmap[pairs1[idx][1]]
			r := rates1[idx]
			dist1[a][b] = maxf(dist1[a][b], r)
			dist1[b][a] = maxf(dist1[b][a], 1.0 / r)
		}
		// FW day1
		for k := 0; k < N; k++ {
			for i := 0; i < N; i++ {
				for j := 0; j < N; j++ {
					if dist1[i][k] > 0 && dist1[k][j] > 0 {
						dist1[i][j] = maxf(dist1[i][j], dist1[i][k] * dist1[k][j])
					}
				}
			}
		}

		// day2
		dist2 := make([][]float64, N)
		for i := 0; i < N; i++ {
			dist2[i] = make([]float64, N)
			for j := 0; j < N; j++ {
				if i == j {
					dist2[i][j] = 1.0
				} else {
					dist2[i][j] = 0.0
				}
			}
		}
		for idx := range pairs2 {
			a := idmap[pairs2[idx][0]]
			b := idmap[pairs2[idx][1]]
			r := rates2[idx]
			dist2[a][b] = maxf(dist2[a][b], r)
			dist2[b][a] = maxf(dist2[b][a], 1.0 / r)
		}
		// FW day2
		for k := 0; k < N; k++ {
			for i := 0; i < N; i++ {
				for j := 0; j < N; j++ {
					if dist2[i][k] > 0 && dist2[k][j] > 0 {
						dist2[i][j] = maxf(dist2[i][j], dist2[i][k] * dist2[k][j])
					}
				}
			}
		}

		maxamt := 1.0
		for i := 0; i < N; i++ {
			amt := dist1[sid][i] * dist2[i][sid]
			if amt > maxamt {
				maxamt = amt
			}
		}
		return maxamt
}
# @lc code=end